### ISSUE TITLE
Support of CMake's configuration of boost

### ISSUE THREAD

**Issue by intractabilis** [mesonbuild/meson#8801]
If you compile boost yourself, a usual way to "compile and forget about it" is to invoke
```
b2 <a lot of options> --stagedir=PREFIX stage
```
It installs the compiled libraries and the `.cmake` files into `PREFIX/lib`. After that it just works with CMake without moving around 15 thousand boost header files.

Meson, unfortunately, explicitly prohibits you from using the official boost `.cmake` configuration for discovering boost dependencies. There is nothing unusual in using the officially supplied by the boost team files. The lack of this support in Meson makes it hard to use with boost.

**Comment by mensinda** [comment]
Is there an actual issue caused by meson not using the `.cmake` files? If you want to use a custom prefix, you can do so by setting `boost_root` in your cross/native file ([see the docs](https://mesonbuild.com/Dependencies.html#boost)).

As for why we don't use the `.cmake` files: The main reason here is that they are a pain to use from meson. Additionally, they also often fail to correctly select the correct combination of library types (see [the boost docs](https://www.boost.org/doc/libs/1_76_0/more/getting_started/windows.html#library-naming) for a complete overview of this mess).

**Comment by intractabilis** [comment]
> Is there an actual issue caused by meson not using the .cmake files?

Of course. Meson has zero idea how to find it without using the installed `.cmake` files. It simply doesn't work in the described scenario.

> If you want to use a custom prefix, you can do so by setting boost_root in your cross/native file (see the docs).

And do it every time for every of 20 projects I downloaded from GitHub?

> see the boost docs for a complete overview of this mess

I am familiar very well with how boost constructs libraries, I've been using it for 15+ years and even contributed a bit. **Boost's generated CMake files work absolutely correctly**. You are reinventing a wheel, while the boost team has already developed the solution.

**Comment by mensinda** [comment]
> Boost's generated CMake files work absolutely correctly.

I have tested it and can confirm that it does **not** work correctly. I can give you a test case in a bit.

> Of course. Meson has zero idea how to find boost without using the installed .cmake files. It simply doesn't work in the described scenario.

First, can we please keep the town a bit more objective? Additionally, could you please provide a way to reproduce your problem? OS, Meson version, boost version, BOOST_ROOT, `meson-logs/meson-log.txt` etc.?

> And do it every time for every of 20 projects I downloaded from GitHub?

Setting the `BOOST_ROOT` environment variable also works if you don't want to bother with meson machine files. Additionally, you can stack machine files, so you can have one global `boost.ini` file with just your BOOST_ROOT.

> Also the prefix doesn't have to be "custom". `/usr/lib` won't work either.

In this case, I really would like to see a test case, because this should work, and does work on my system and our CI.

> You are reinventing a wheel, while the boost team has already developed the solution.

I would agree if the solution would actually work *and* if it would be easy to use for meson. I initially also tried to use CMake, when I refactored the previous implementation of our boost system dependency. However, this attempt failed (see #6569). Here are some of the issues I ran into (that I can still remember):

- Issues with CMake not selecting the correct boost prefix when more than one boost installation is present
- Requires a custom `CMakeLists.txt` for handling special boost logic
- Basically also requires a custom backend in meson for aggregating all the CMake results
- Mapping the existing boost module strings to CMake targets is a pain

Of course, all of these points can be solved, however, at some point "reinventing" the wheel was easier than using a partially broken wheel that was not designed to work with Meson.

So "just plug in CMake and it works" is definitely not the case here and will involve a major amount of effort while also introducing another point of failure with CMake. Thus, the current priority is to improving the current implementation.

**Comment by intractabilis** [comment]
> I have tested it and can confirm that it does not work correctly. I can give you a test case in a bit.

Alright. Maybe it has failing test cases. By ignoring the official files supported by the boost development team Meson does not and will not do a better job.

> Setting the BOOST_ROOT environment variable also works if you don't want to bother with meson machine files. Additionally, you can stack machine files, so you can have one global boost.ini file with just your BOOST_ROOT.

Jussi had a long rant about how bad environment variables are. I can come up with million ways for workaround, starting from not using Meson. However, Jussi's words were that there should be one big button on the table with a word "Build" on it.

> In this case, I really would like to see a test case

Gosh. For boost `.cmake` files is the only right source of the location of the header files. Always, in any scenario. BOOST_ROOT is a fossil left from 20+ years of history. If in your custom written script you find it in some other way, you are wrong.

> Thus, the current priority is to improving the current implementation.

How you are going to do it if the headers location is in `.cmake` files?

**Comment by intractabilis** [comment]
Another way could be is to work with the Boost team to provide something specific for Meson.

**Comment by mensinda** [comment]
First of all, my point with `BOOST_ROOT` was to give an alternative and not to recommend it over machine files. Also, our goal **is** to just make it work out of the box. However, I can't help you / fix the bug if you don't provide more information to reproduce the problem.

> Gosh. For boost .cmake files is the only right source of the location of the header files. 

I get your point with `.cmake` files being the official source of truth for Boost. However, just plugging in CMake has also major issues (see my previous post) and is also a massive maintenance burden. So we are not going to switch unless there is a very good reason. Specifically, for us to switch to CMake you have to argue how it would be far more work for us to maintain our current implementation by adding support for more edge cases, than trying to implement a CMake backend.

Please do remember that I actually tried to use CMake initially but gave up because this setup was far too error-prone in other ways.

> Another way could be is to work with the Boost team to provide something specific for Meson.

This would be a great solution, however, something like this shouldn't be Boost specific and a more general format (we are not going to ask the boost devs to implement a specific format just for Meson that is only understood by Meson).

**Comment by intractabilis** [comment]
>However, I can't help you / fix the bug if you don't provide more information to reproduce the problem.

All information is in my initial post. When you compile boost, it generates `.cmake` files with the right location of the header files, so you don't have to copy 15 thousand (not an exaggeration) files to your PREFIX. Without analyzing the `.cmake` files Meson has no idea where the headers are located. I am not sure what else you need to reproduce. Copy boost to folder1, run Meson in folder2. How Meson will find boost in folder1 without any information? The information is in your PREFIX/lib inside `.cmake` files.

**Comment by mensinda** [comment]
Wait, so does `stage` only install the `.cmake` files, which then reference the libraries in the build dir and includes from the source dir?

**Comment by intractabilis** [comment]
It installs the libraries and `.cmake` files to PREFIX/lib. That's what I said in my initial post. `.cmake` files have the right location of the header files, so you don't have to copy 15 thousand header files to every PREFIX you have. The compiled libraries and the `.cmake` files on other hand are in the standard location inside PREFIX.

**Comment by intractabilis** [comment]
Ah... `.cmake` files of course also have the right names of the libraries, eleminating completely the naming mess.

**Comment by mensinda** [comment]
OK, sorry I misunderstood your setup. I was assuming that all files (including headers) were installed and Meson was unable to find the installed headers, while the `.cmake` files had the correct paths.

I have also installed a stage prefix now and see the problem...

Unfortunately, there is currently no automatic solution for this use case. The only workaround that is currently available is to manually set `boost_include` and `boost_librarydir` in a Meson machine file (which I admit is less comfortable than what CMake currently offers).

Just to make sure we are talking about the same thing: When I talk about a CMake backend, I mean that Meson generates a small `CMakeLists.txt` and actually runs CMake to extract the correct paths from the CMake trace output. We are *not* parsing `.cmake` files manually.

However, for this specific use case it *may* be possible to grep for `_BOOST_INCLUDEDIR` in the `.cmake` files (if present) and try to use the result. But again, adding a full CMake backend for this use case will take a lot of work (with no guarantee of success), which is hard to justify if there is a workaround...

> Ah... .cmake files of course also have the right names of the libraries, eliminating completely the naming mess.

The main problem with the library names is that the mapping from the library name () to the CMake target (`filesystem` --> `Boost::filesystem`) is not always trivial (especially for the python related libraries + supporting legacy library names in meson).

**Comment by intractabilis** [comment]
Great, I ma glad a misunderstanding was resolved.

The purpose of creating this issue wasn't to help me to find a workaround. That I can do myself. I was just trying to point to the right thing to do for the benefit of Meson. How hard to implement the right thing, or if it will be implemented, because it's hard, is another issue.

Generating a `CMakeLists.txt` and parse the `cmake`'s output sounds like a solid approach. Why not to do it for boost?

**Comment by intractabilis** [comment]
Btw, I agree that creating a full CMake back-end is a non-starter. It will be a nightmare, because of different version, bugs in CMake, various policies, etc.

**Comment by mensinda** [comment]
> Generating a `CMakeLists.txt` and parse the cmake's output sounds like a solid approach. Why not to do it for boost?

This is how we handle all our CMake dependencies and it works great for 99.9% of them. For all dependencies that don't provide `*.pc` files and only `*.cmake` files this approach is used.

This base approach was used in #6569. Sadly, Boost is in the 0.1% of dependencies where this approach was too painful to get working.

**Comment by mensinda** [comment]
Essentially, almost all CMake dependencies work just fine with our generic `CMakeLists.txt`. However, to get boost working we need a more specialized `CMakeLists.txt` with additional boost-specific logic. Specifically, getting this last part to work reliably was the main issue here.

**Comment by intractabilis** [comment]
It reads that you gave up because CMake cannot find boost. If CMake cannot find boost, it's definitely not a Meson's problem. It's either CMake's or the authors of the boost's `.cmake` files.

**Comment by intractabilis** [comment]
> Issues with CMake not selecting the correct boost prefix when more than one boost installation is present

This is not supposed to work. Boost's `.cmake` files are specifically generated for a particular installation.

**Comment by mensinda** [comment]
If I remember correctly (it has been a while since then), the main problem was that CMake found the *wrong* boost libraries. So yeah, not our problem. Thus, the solution for Meson was to implement a custom lookup mechanism that worked in these cases.

> This is not supposed to work. Boost's .cmake files are specifically generated for a particular installation.

I think here the problem was that one installation had the Boost `.cmake` files and one didn't and the `FindBoost.cmake` from CMake itself was used here.

Anyway, I will look into manually parsing `_BOOST_INCLUDEDIR` as a fallback when there are no Boost headers in the "normal" search paths.

**Comment by intractabilis** [comment]
Libraries are supposed to be in the PREFIX/lib location, CMake doesn't "search" for them.  It sounds like a weird messed up case.

I would just restore that solution you thought didn't work and add it as an option when a user specifically requests `method : 'cmake'`.

**Comment by vinipsmaker** [comment]
> Is there an actual issue caused by meson not using the `.cmake` files?

For one, I cannot use Boost packages installed by vcpkg.

**Comment by alex-tee** [comment]
Just stumbled upon this after installing boost from vcpkg and attempting to use it via meson. Is there still no way to make meson find boost when installed via vcpkg?

**Comment by intractabilis** [comment]
The only way to reliably work with boost in Meson is to add the following lines to your native/cross-build file:
```
[properties]
boost_includedir = '/somepath/boost_1_86_0'
boost_librarydir = '/couldbeanotherpath/lib'
```

**Comment by alex-tee** [comment]
This means you can't use the vcpkg version, you have to build boost yourself. For me, that's ok because i'm setting it up on my own CI server, but not being able to use an existing installation from vcpkg seems inconvenient for a lot of people.

**Comment by intractabilis** [comment]
Yeah... This is not a particularly pretty part of Meson. I am curious how difficult would it be to disable a special treatment of boost in the Python source code of Meson, so the boost configuration relies on CMake files like any other CMake-based package. Maybe it will work without any customization whatsoever.
